src/ClaimToken.sol

```diff
@@ -0,0 +1,202 @@
+        emit EventCreated(tokenAddress, eventID);
+    }
+
+    // Updates the status of an existing event
+    function updateEvent(address tokenAddress, string calldata eventID, bool isEventClosed)
+        external
+        override
+        onlyOwner
+    {
+        bytes32 eventIDHash = _hashString(eventID);
+
+        require(_isEventOngoing[tokenAddress][eventIDHash] != !isEventClosed, "Event already in this state");
+
+        _isEventOngoing[tokenAddress][eventIDHash] = !isEventClosed;
+        emit EventUpdated(tokenAddress, eventID, isEventClosed);
+    }
+
+    // Claims tokens for a user based on the provided data
+    function claim(
+        address tokenAddress,
+        string calldata eventID,
+        address userAddress,
+        uint256 amount,
+        bytes calldata signerSignature
+    ) external override nonReentrant {
+        bytes32 eventIDHash = _hashString(eventID);
+        require(_isEventOngoing[tokenAddress][eventIDHash], "Event is closed");
+
+        require(getTokenBalance(tokenAddress) >= amount, "Insufficient balance");
+
+        bytes32 claimHash = keccak256(abi.encode(tokenAddress, eventIDHash, userAddress, amount));
+        bytes32 ethSignedMessageHash = MessageHashUtils.toEthSignedMessageHash(keccak256(abi.encode(claimHash)));
+        address signer = ECDSA.recover(ethSignedMessageHash, signerSignature);
+
+        require(_isActivatedSigner[signer], "Invalid signer");
+
+        require(IERC20(tokenAddress).transfer(userAddress, amount), "Token transfer failed");
+
+        _userClaimedAmount[tokenAddress][eventIDHash][userAddress] += amount;
+
+        emit Claimed(claimHash, tokenAddress, eventID, userAddress, amount, signerSignature);
+    }
+
+    // Transfers tokens from the contract to a recipient
+    function transferToken(address tokenAddress, address recipient, uint256 amount) external onlyOwner {
+        require(tokenAddress != address(0), "Invalid token address");
+        require(recipient != address(0), "Invalid recipient address");
+        require(getTokenBalance(tokenAddress) >= amount, "Insufficient balance");
+
+        IERC20(tokenAddress).transfer(recipient, amount);
+    }
+
+    // ------------------------
+    // -- Internal Functions --
+    // ------------------------
+
+    function _hashString(string memory input) internal pure returns (bytes32) {
+        return keccak256(abi.encodePacked(input));
+    }
+
+    // -----------------------
+    // -- Private Functions --
+    // -----------------------
+
+    // Updates signers and their activation status
+    function _updateSigners(address[] memory _signerList, bool[] memory _isActivatedList) private {
+        require(_signerList.length == _isActivatedList.length, "Mismatch in input lengths");
+
+        for (uint256 i = 0; i < _signerList.length; i++) {
+            address signer = _signerList[i];
+            bool isActivatedSigner = _isActivatedList[i];
+            _isActivatedSigner[signer] = isActivatedSigner;
```

Comment:
We can check that we are not activated an already activated signer or deactivate an already deactivated signer.